python/ray/autoscaler/v2/instance_manager/subscribers/reconciler.py

```diff
@@ -0,0 +1,101 @@
+import logging
+import threading
+from concurrent.futures import ThreadPoolExecutor
+from typing import List
+
+from ray.autoscaler.v2.instance_manager.instance_storage import (
+    InstanceStorage,
+    InstanceUpdatedSuscriber,
+    InstanceUpdateEvent,
+)
+from ray.autoscaler.v2.instance_manager.node_provider import NodeProvider
+from ray.core.generated.instance_manager_pb2 import Instance
+
+logger = logging.getLogger(__name__)
+
+
+class InstanceReconciler(InstanceUpdatedSuscriber):
+    """InstanceReconciler is responsible for reconciling the difference between
+    node provider and instance storage. It is also responsible for handling
+    failures.
+    """
+
+    def __init__(
+        self,
+        instance_storage: InstanceStorage,
+        node_provider: NodeProvider,
+        reconciler_interval_s: int = 120,
+    ) -> None:
+        self._instance_storage = instance_storage
+        self._node_provider = node_provider
+        self._failure_handling_executor = ThreadPoolExecutor(max_workers=1)
+        self._reconcile_timer = threading.Timer(
```

Comment:
should this be a periodic thing rather than just a delayed timer?

Reply:
and should this routine be called in a separate thread pool also?

Reply:
this is called periodically in a different thread according to threading.Timer api